### PR TITLE
Remove Dependency to `memoise` and `digest`

### PR DESCRIPTION
**Pull request by brodieG**
### TL;DR

* found funny bug involving `memoise` and the `crayon` CRAN *binary* package
* realized you can remove dependency to `memoise` and `digest` with almost no additional code by using a closure
* created this pull request that removes that dependency

@jimhester, I'm guessing this issue is of interest to you given the link to `memoise`.  Also, I apologize for lobbying for the removal of `memoise` as an import for `crayon` and hope you find my request is not completely crazy even if you don't agree with it.

### Whole Story

I ran into a rather bizarre bug involving `memoise` that made it impossible to `forget` the cached version of `crayon:::i_num_colors`.  Somehow, the binary version of `crayon` on CRAN has a corrupted copy of the memoised `crayon:::i_num_colors` function:

    ls(all=T)
    # character(0)
    install.packages("crayon")
    # trying URL 'http://cran.r-project.org/bin/macosx/mavericks/contrib/3.2/crayon_1.3.1.tgz'
    # Content type 'application/x-gzip' length 604079 bytes (589 KB)
    # ... snip
    str(crayon:::i_num_colors)
    # function (...)  
    #  - attr(*, "memoised")= logi TRUE

Notice how the function does not have a memoise class as it should ([full session transcript on gist](https://gist.github.com/brodieG/6101ada64d917b4b659d)).  This causes `is.memoised` to fail, which then prevents `forget` from working.

Installing from source everything works fine:

    remove.packages("crayon")
    q()  # and restart R
    install.packages("crayon", type="source")
    str(crayon:::i_num_colors)
    # function ()  
    #  - attr(*, "memoised")= logi TRUE
    #  - attr(*, "class")= chr [1:2] "memoised" "function"

I also rebuilt the package binary and installed it, and everything was fine, so something funny happened with the CRAN binary build process.

All that aside, while debugging this problem I realized that the `memoise` and `digest` dependencies of your package hinge only on memoising `crayon:::i_num_colors`.  Since that function doesn't even take arguments I think it makes sense to use a closure to handle the memoization rather than add dependencies to a lot of functionality that is not really needed (granted, as far as adding dependencies `memoise` and `digest` are probably as good as it gets).

I think your package is awesome and should be used by anyone writing packages producing meaningful amounts of text output.  I initially balked at the idea of importing your package into mine due to my desire to keep my package download small and as such minimize the dependency footprint, and that all I needed was to color some stuff red or green.  In fact, I went so far as to borrow a subset of the `crayon` functionality and put it into another package without dependencies until I realized how silly that was (which btw was around the time I realized I needed to `substr` colored strings and discovered you'd already handled all the painful stuff you need to do with closing and re-opening escape sequences).

I'm going to import your package no matter what, but hopefully you'll consider removing the dependencies to `memoise` / `digest` to make this package stand alone, especially since the change required to do so is so minor.

### Other Notes

I added some tests, and in the process of re-running tests and checking the package I ran into a couple of issues:

* I had to change how options were being re-set in `test-has-colors.R`
* I had to remove references to `%:%`; I have no idea where to find that function and it doesn't seem to be part of this package or any of the dependencies I removed, and google is totally useless it telling me where it comes from.  I inferred that it is a safe version of `:` (i.e. won't fail with `1:x` where `x==0`) and replaced it accordingly
* I had to add `:::` references to non-exported variables in `test-make-style`

### CODE REVIEW

**Comment by gaborcsardi** [comment]
> Notice how the function does not have a memoise class as it should (full session transcript on gist). This causes is.memoised to fail, which then prevents forget from working.

I think that's because the `memoise` class is a new thing in the `memoise` package, and the `crayon` binary was built with an old version. So an easy workaround would be to ask CRAN to rebuild `crayon`.

OTH, I very much favour of removing dependencies in general, given the rigid dependency mechanism of R packages.

I do have some questions about the PR, but I guess I can fix these after merging.

Thanks much!

File: tests/testthat/test-operations.R [hunk 1]
```diff
@@ -47,7 +50,7 @@ test_that("col_substr, start after string end", {
   expect_equal(col_substr("red", 3, 4), "d")
   expect_equal(col_substr("red", 3, 5), "d")
   expect_equal(strip_style(col_substr("\033[31mred\033[39m", 3, 4)), "d")
-  expect_equal(strip_style(col_substr("\033[31mred\033[39m", 3, 5)), "d")  
+  expect_equal(strip_style(col_substr("\033[31mred\033[39m", 3, 5)), "d")
```
Comment (gaborcsardi) [inline tests/testthat/test-operations.R]: Could you please not change the whitespace? I know that trailing whitespace is annoying, but whitespace changes are annoying, too. :)

File: tests/testthat/test-make-style.r [hunk 2]
```diff
@@ -30,13 +30,13 @@ test_that("hexa color regex works", {
                 "foo#123456bar")
 
   for (color in positive) {
-    expect_true(grepl(hash_color_regex, color))
-    expect_true(grepl(hash_color_regex, toupper(color)))
+    expect_true(grepl(crayon:::hash_color_regex, color))
+    expect_true(grepl(crayon:::hash_color_regex, toupper(color)))
```
Comment (gaborcsardi) [inline tests/testthat/test-make-style.r]: I think this is fine without `:::` as long as you run the tests via `testthat`? No?

File: DESCRIPTION [hunk 3]
```diff
@@ -1,6 +1,6 @@
 Package: crayon
 Title: Colored Terminal Output
-Version: 1.3.1
+Version: 1.3.2
 Authors@R: person("Gabor", "Csardi", , "csardi.gabor@gmail.com", role = c("aut", "cre"))
```
Comment (gaborcsardi) [inline DESCRIPTION]: Can you please add yourself to the authors, and also a bullet point to the NEWS? Thanks.

**Comment by gaborcsardi** [comment]
I made some comments. Do you want to resubmit the PR? I can also just merge it as it is, and then make some changes....

**Comment by brodieG** [comment]
Thanks for considering the PR.  I'll address your comments and resubmit hopefully at some point today.

**Comment by jimhester** [comment]
You don't need to resubmit, just push the changes to the same branch. If
you don't want the extra commits squash them.

File: tests/testthat/test-make-style.r (hunk 2 above)
Comment (brodieG) [inline tests/testthat/test-make-style.r]: Yes, indeed.  I am in the habit of using `test_dir` and wasn't thinking it wouldn't affect your tests as run with `test_check`.

Comment (brodieG) [inline tests/testthat/test-make-style.r]: Actually, this is also the same thing that tripped me up with `%:%`, not to mention I apparently don't know how to grep on the command line.

**Comment by gaborcsardi** [comment]
@brodieG Btw. if you re-push your PR branch, then please write a comment here, because I don't get a notification for the push, only for comments here. Thanks!

File: DESCRIPTION (hunk 3 above)
Comment (brodieG) [inline DESCRIPTION]: Just noticed you have a NEWS file at top level and in `inst`.  Should I update both?

Comment (gaborcsardi) [inline DESCRIPTION]: One is a symlink to the other I believe.

**Comment by brodieG** [comment]
Hopefully this addresses your comments.  Thanks for your patience.

**Comment by gaborcsardi** [comment]
Great, thanks!
